graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java

```diff
@@ -56,42 +57,60 @@ public EntitySuggestionResponse suggest(final String collection,
                                             final int page,
                                             final int perPage,
                                             final Subject subject) {
-
-        if (!hasAllPermission(subject) && !hasReadPermissionForWholeCollection(subject, collection)) {
-            return new EntitySuggestionResponse(List.of(),
-                    PaginatedList.PaginationInfo.create(0, 0, page, perPage));
-        }
-
         final MongoCollection<Document> mongoCollection = mongoConnection.getMongoDatabase().getCollection(collection);
 
-        Bson bsonFilter = (query != null && !query.isEmpty()) ?
-                Filters.regex(valueColumn, query, "i") :
-                new BsonDocument();
+        final var bsonFilter = !Strings.isNullOrEmpty(query)
+                ? Filters.regex(valueColumn, query, "i")
+                : Filters.empty();
 
-        final FindIterable<Document> documents = mongoCollection
+        final var resultWithoutPagination = mongoCollection
                 .find(bsonFilter)
                 .projection(Projections.include(valueColumn))
-                .sort(Sorts.ascending(valueColumn))
-                .limit(perPage)
-                .skip((page - 1) * perPage);
+                .sort(Sorts.ascending(valueColumn));
+
+        final var userCanReadAllEntities = hasAllPermission(subject) || hasReadPermissionForWholeCollection(subject, collection);
+        final var skip = (page - 1) * perPage;
+        final var checkPermission = createPermissionCheck(subject, collection);
```

Comment:
It is a premature call - this creation is only needed for "I do not have access to all entities" path.

Reply:
Construction of it is very cheap, I am constructing it here so I can reuse it later in the counts query.